mantidproject/mantid
# FFT automatic phase correction

Added an _AutoShift_ property to `FFT` at the request of scientists.

`FFT` assumes that the input workspace's X data is centred at zero.
If this is not true, the output will have the wrong phase.
This can be worked around by using the property _Shift_, but the requirement was to do this automatically.
- Added _AutoShift_ property, which automatically calculates and applies the correct shift
- Added unit tests and extended system test
- Updated algorithm documentation

**To test:**

_Code review:_
In particular, the changes to the documentation and the name of the new parameter.
I don't like the name "AutoShift" but couldn't think of anything better to call it - suggestions welcome!

_Functional tests:_
The new tests should cover most of this.
Using the script from the issue:

```
from math import pi, sin, cos
X = []
YRe = []
YIm = []
E = []
N = 1000
w = 10
for i in range(0,N):
       x = 2*pi*i/N
       X.append(x)
       YRe.append(cos(w*x))
       YIm.append(sin(w*x))
       E.append(0.1)
ws = CreateWorkspace(DataX=X+X,DataY=YRe+YIm,DataE=E+E,NSpec=2)

FFT(InputWorkspace='ws', OutputWorkspace='FFT1', Imaginary=1)

factor = pi/w
ScaleX(InputWorkspace='ws', OutputWorkspace='Offsetted', Factor=factor, Operation='Add')
FFT(InputWorkspace='Offsetted', OutputWorkspace='FFT2', Imaginary=1)
FFT(InputWorkspace='Offsetted', OutputWorkspace='FFT3', Imaginary=1, AutoShift=True)
```

The transforms `FFT1` and `FFT2` should compare equal (wrong) but `FFT1` and `FFT3` should have a phase difference.

Fixes #15279

Release notes have been updated.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
